servicex_local/science_images.py

```diff
@@ -381,3 +380,137 @@ def transform(
+        self,
+        generated_files_dir: Path,
+        input_files: List[str],
+        output_directory: Path,
+        output_format: str,
+    ) -> List[Path]:
+        """Transform the input files and return the path to the output file.
+
+        This method works by invoking the container with a specific transformation command.
+
+        Args:
+            generated_files_dir (str): Directory for generated files
+            input_files (List[str]): List of input files
+            output_directory (Path): The output directory
+            output_format (str): The desired output format
+
+        Returns:
+            List[Path]: List of output file paths
+        """
+
+        output_paths = []
+        x509up_path = Path(os.getenv("TEMP", "/tmp")) / "x509up"
+        if x509up_path.exists():
+            x509up_volume = ["--bind", f"{x509up_path}:/tmp/grid-security/x509up"]
+        else:
+            logger = logging.getLogger(__name__)
+            logger.warning("x509up certificate not found at /tmp/x509up")
+            x509up_volume = []
+
+        for input_file in input_files:
+            safe_image = self.image_uri.replace(":", "_").replace("/", "_")
+            container_name = f"sx_codegen_container_{safe_image}_{Path(input_file).stem}"
+
+            output_name = Path(input_file).name
+
+            if input_file.startswith(("root://", "http://", "https://")):
+                input_volume = []
+                container_path = input_file
+            else:
+                input_path = Path(input_file)
+                if not input_path.exists():
+                    raise FileNotFoundError(
+                        f"Input file for Singularity image {input_file} not found."
+                    )
+                input_volume = ["--bind", f"{str(input_path.absolute())}:/input_file.root"]
+                container_path = "/input_file.root"
+
+            file_runner = """#!/bin/bash
+python_cmd=$(command -v python3 || command -v python)
+exec $python_cmd /generated/kick_off.py $@
+"""
+            with open(generated_files_dir / "file_runner.sh", "w", newline="\n") as f:
+                for ln in file_runner.splitlines():
+                    f.write(ln.strip() + "\n")
+
+            kick_off = """
+import json
+import os
+import sys
+
+x509up_path = "/tmp/grid-security/x509up"
+if os.path.exists(x509up_path):
+    os.chmod(x509up_path, 0o600)
+
+with open("/generated/transformer_capabilities.json") as f:
+    info = json.load(f)
+file_to_run = info["command"]
+arg1 = sys.argv[1]
+arg2 = sys.argv[2]
+arg3 = sys.argv[3]
+if info["language"] == "python":
+    exe = sys.executable
+    ret_code = os.system(exe + " " + file_to_run + " " + arg1 + " " + arg2 + " " + arg3)
+elif info["language"] == "bash":
+    ret_code = os.system("bash " + file_to_run + " " + arg1 + " " + arg2 + " " + arg3)
+else:
+    raise ValueError("Unsupported language: " + info["language"])
+exit_code = ret_code >> 8
+sys.exit(exit_code)
+"""
```

Comment:
If the above code is identical to the above, we should factor it out into a separate, common, method.

Reply:
This is true for all this code in here. Perhaps there is a nice way to make the docker/singularity code unique, and the common code... common.